ot_flash: Add HW operation support & improved memory protection

AlexJones0 opened the pull request:
This PR is the second of a series of 3 PRs to improve the OpenTitan Flash Controller emulation and make it more feature complete, with a focus on the features that are needed to support keymgr integration.

This PR focuses on refactoring and introducing additional logic to allow the flash controller to support both software- and hardware-initiated flash operations. For context, SW and HW operations use the same controller operation logic, but:
1. HW operations are not visible to SW registers, except for fatal faults, REGWEN locking, etc.
2. SW and HW operations share one program FIFO but have separate read FIFOs
3. SW and HW operations have completely different data & info page protection mechanisms, where protection is based on the current flash controller's lifecycle state.

As such, this PR:
 - Implements some stubbed flash lifecycle management (support for initialization including loading flash secrets, but not for RMA entry).
 - Implements generalized support for HW flash operations.
 - Implements HW data page and info page protection mechanisms.
 - Implements additional lc_ctrl-backed secret protection (qualifications), though with no lc_ctrl connections in this PR.

See the commit messages for more details about each change.

Note also that the HW operations are currently unused in this PR, but will be utilized in a subsequent PR that will add an interface for retrieving keymgr secrets from the flash, and connect the flash to the lc_ctrl.

AlexJones0 committed:
[ot] hw/opentitan: ot_flash: Centralize fifo status & interrupt updates

Refactor FIFO status updates to all occur after finishing the current
execution of an operation, rather than handling all the individual
status and interrupt logic.

This is equivalent and easier to understand/maintain, and will
importantly allow easier implementation of hardware flash controller
reads in the future.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 committed:
[ot] hw/opentitan: ot_flash: Add initial flash lcmgr state

Add some basic flash lifecycle manager phases which are used for RMA
entry and initialization, starting in the `None` phase and moving to
the `Seed` phase during initialization.

Also take the opportunity to re-order the flash traces whilst adding a
new one as clean-up.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 committed:
[ot] hw/opentitan: ot_flash: Initial HW request support

Hardware flash requests are processed in the same way as flash
controller requests, but have key differences in how they update the
state of software-visible status registers and behave on errors (they
fault with a fatal error that needs a reset).

Hardware operation requests also have different protection rules
depending on the lifecycle phase, but that is stubbed as unimplemented
for now (all HW-sourced operations are allowed).

Hardware and software operations share a program FIFO, but do not share
read buffers. To facilitate sharing a FIFO, software is locked out of
writing to the program FIFO while there is not an active *sw-initiated*
program operation.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 committed:
[ot] hw/opentitan: ot_flash: Add HW request data page permissions

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 committed:
[ot] hw/opentitan: ot_flash: Add SW info page qualification checks

Some info pages that hold secrets (e.g. the creator and owner seeds for
key derivation by the keymgr, the isolated partition for provisioning,
etc.) are locked by an additional layer of protections on top of the
existing info page configuration registers, which depend on signals
broadcast by the lifecycle controller.

Since the lifecycle controller is not connected, these signals are
stubbed for now, but the logic for calculating the different permissions
on these pages and merging them with the existing sw register config is
all implemented to allow correct info page protection for software
operations.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 committed:
[ot] hw/opentitan: ot_flash: Add HW request info page permissions

As with data pages, hardware requests bypass the sw-register-configured
protections and instead have a set of hardware protection rules applied
to them. The list of rules is searched for any matches (based on both
the specific info page and the current flash lcmgr phase), allowing
hardware only selective access to specific info pages containing secrets
during certain life cycle phases.

For these regions, the ECC and scramble properties can be optionally
disabled via the HW_INFO_CFG_OVERRIDE register to aid in recovery. This
is implemented as well. There are also OTP secrets that configure the
same properties, but these are left unimplemented for now.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 committed:
[ot] hw/opentitan: ot_flash: Gate controller ops on initialization

This commit gates controller operations on initialization. It allows
writes to the `CONTROL` register, but the actual operation does not
start until the flash controller is initialized.

By default, the flash controller arbiter does not allow either SW or HW
requests. When the `INIT` command is sent and the flash physical
controller (around the flash macro) is initialized, the arbiter then
immediately switches to prioritising hardware to service the initial
hardware read requests to read the keymgr seeds out of the flash, and
then waits for software requests. Since reading the keymgr seeds is the
last part of the flash controller lcmgr initialization, this means that
when the flash controller is initialized, we can process a software op.

Specifically, if the `CONTROL` register was already written with
`start=true` before/while initializing, then this will cause the
operation to begin *as soon as initialization is finished* (and the
flash arbiter switches to prioritising SW requests over HW). Thus,
we must also begin an operation in the case that `CONTROL.START=true`
at the point in time where flash controller initialization finishes.

As well as moving the `CONTROL` processing logic into its own function,
this commit refactors the code slightly.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 committed:
[ot] hw/opentitan: ot_flash: Clear program FIFO on HW start / SW finish

As part of the logic for synchronizing the shared use of the program
FIFO between hardware and software flash operations (in addition to the
existing locking of SW writes to the PROG_FIFO), we always clear the
program FIFO at the start of a hardware operation, and clear any data
left over in the program FIFO at the end of a SW operation, regardless
of which operation it is.

To be sure that we only clear the program FIFO once when starting a HW
operation, create a new `ot_flash_op_start` function for starting an
operation, whereas `ot_flash_op_execute` can instead be used to continue
executing a started operation.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 committed:
[ot] hw/riscv: ot_earlgrey: Format header includes

This file was not fully formatted properly, run clang-format to re-order
and use the correct include orders.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

rivos-eblot commented on hw/opentitan/ot_flash.c:
```diff
@@ -769,16 +809,103 @@ struct OtFlashState {
 
+/* Flash memory protection rules */
+
+static const OtFlashPropertyCfg OtFlashCfgAllowRead = {
```
nit pick: constant names need to be fully upper-cased.

rivos-eblot commented on hw/opentitan/ot_flash.c:
```diff
@@ -1013,31 +1150,120 @@ static uint32_t ot_flash_get_info_page_cfg_reg(
-static bool
-ot_flash_info_page_cfg_op_enabled(OtFlashState *s, uint32_t info_page_cfg_reg)
+static OtFlashPropertyCfg ot_flash_get_hw_info_page_cfg(
+    OtFlashState *s, unsigned bank, unsigned info_partition, unsigned page)
```
`s` seems to be const

AlexJones0 replied:
Good spot, I'm often missing these.

rivos-eblot commented on hw/opentitan/ot_flash.c:
```diff
@@ -1013,31 +1150,120 @@ static uint32_t ot_flash_get_info_page_cfg_reg(
+}
+
+static OtFlashPropertyCfg
+ot_flash_get_info_page_reg_cfg(OtFlashState *s, uint32_t info_page_cfg_reg)
```
maybe `s` const as well

rivos-eblot commented on hw/opentitan/trace-events:
```diff
@@ -196,25 +196,27 @@ ot_entropy_src_update_filler(bool iok, bool ook, bool pok, bool all) "in %u, out
 ot_flash_prog_word(bool info_part, unsigned word_addr, uint32_t word) "info_part %u word_addr %u word %u"
-ot_flash_erase(bool info_part, unsigned word_addr, unsigned num_bytes) "info_part %u word_addr %u num_bytes %u"
+ot_flash_reset_fifo(const char *name) "%s"
+ot_flash_set_error(const char *op, bool hw, uint32_t err_code, uint32_t err_addr) "%s (hw=%u): err=%08x at addr=0x%06x"
```
QEMU checkpatch (or some other tool, can't remember) that "%08x" is not prefixed with `0x`

AlexJones0 replied:
Wow thanks that is very useful, I didn't know about that. I can see some checkpatch violations in the existing codebase, so I can try and fix things as I touch them.

AlexJones0 replied:
Also, checkpatch and `clang-format.yml` were disagreeing about spacing around colons in bitfields, so I've added a commit to update `clang-format.yml` with this change to align the two.

rivos-eblot commented on hw/opentitan/ot_flash.c:
```diff
@@ -1589,6 +1895,99 @@ static bool ot_flash_check_program_type(OtFlashState *s)
+    switch (op) {
+    case CONTROL_OP_READ:
+        s->op.kind = OP_READ;
+        s->op.address = s->regs[R_ADDR] & ~3u;
```
might be worth defining a constant/macro for this recurrent masking

jwnrt commented on hw/opentitan/ot_flash.c:
```diff
@@ -730,6 +733,25 @@ typedef struct {
     unsigned size; /* size in bytes of the partition */
 } OtFlashInfoPart;
 
+typedef enum {
```
Could you add a comment saying something like "page 1 is only accessible with the creator qualification, page 2 with owner, etc"

AlexJones0 replied:
I've left more thorough documentation on the `ot_flash_update_info_page_qualification` function and just added a small sentence describing qualifications here with a pointer to that documentation.

jwnrt commented on hw/opentitan/ot_flash.c:
```diff
@@ -1057,31 +1079,107 @@ static uint32_t ot_flash_get_info_page_cfg_reg(
+        info_partition != FLASH_SEED_INFO_PARTITION) {
+        return;
+    }
+    switch (info_partition) {
```
This should be `switch(page)`

jwnrt commented on hw/opentitan/ot_flash.c:
```diff
@@ -1057,31 +1079,107 @@ static uint32_t ot_flash_get_info_page_cfg_reg(
+}
+
+static void
+ot_flash_update_info_page_qualification(unsigned bank, unsigned info_partition,
+                                        unsigned page, OtFlashPropertyCfg *cfg)
```
Not sure if this is QEMU style but can you document that `*cfg` is an output?

jwnrt commented on hw/opentitan/ot_flash.c:
```diff
@@ -1057,31 +1079,107 @@ static uint32_t ot_flash_get_info_page_cfg_reg(
+}
+
+static void
+ot_flash_update_info_page_qualification(unsigned bank, unsigned info_partition,
```
Qualifications are underdocumented on the OpenTitan side, could you add a comment describing them to this function?

Link to https://opentitan.org/book/hw/top_earlgrey/ip_autogen/flash_ctrl/index.html#isolated-information-partitions

rivos-eblot reviewed (commented):
I read nothing suspicious however I've forgotten how the flash controller works, so leaving to people with better knowledge the approval for these changes. Seems ok to me.

jwnrt commented on hw/opentitan/ot_flash.c:
```diff
@@ -730,6 +733,25 @@ typedef struct {
+} OtFlashQualifiedInfoPage;
+
+typedef union {
+    uint8_t bitmap;
+    struct {
+        uint8_t en : 1;
+        uint8_t rd_en : 1;
+        uint8_t prog_en : 1;
+        uint8_t erase_en : 1;
+        uint8_t scramble_en : 1;
+        uint8_t ecc_en : 1;
+        uint8_t he_en : 1;
+    };
+} OtFlashPropertyCfg;
```
I haven't used C bitfields much but `cppreference` says the type should be `unsigned int`, `signed int` or `int`, is `uint8_t` okay here?

https://en.cppreference.com/w/c/language/bit_field.html

@rivos-eblot this seems to be implementation-defined, do you have a preference?

> Additional implementation-defined types may be acceptable. It is also implementation-defined whether a bit-field may have [atomic](https://en.cppreference.com/w/c/language/atomic.html) type.(since C11) The number of bits in a bit-field (width) sets the limit to the range of values it can hold:

rivos-eblot replied:
You are right, we should stick with unsigned.

AlexJones0 replied:
Yes, this makes sense - there is a couple of existing instances of this in the QEMU code (`OtPwrMgrEvents` in the `ot_pwrmgr.c`, and also `OtOTPPartDesc` in `ot_otp_eg.c` and `ot_otp_dj.c`) which would be good to fix at some point separate from this PR.

jwnrt commented on hw/opentitan/ot_flash.c:
```diff
@@ -730,6 +733,25 @@ typedef struct {
+    FLASH_QUAL_INFO_PAGE_ISOLATED = 3,
+} OtFlashQualifiedInfoPage;
+
+typedef union {
```
I don't know if this is important, but could you add a `static_assert(sizeof(OtFlashPropertyCfg) == 1)` to make sure the flags don't outgrow the `uint8_t`?

AlexJones0 replied:
Sure, but with the change to `unsigned` as above I now check `sizeof(OtFlashPropertyCfg) == sizeof(unsigned int)`.

jwnrt commented on hw/opentitan/ot_flash.c:
```diff
@@ -1304,30 +1395,45 @@ static unsigned ot_flash_next_info_address(OtFlashState *s)
+                                      ECC_DIS) == OT_MULTIBITBOOL4_TRUE;
+        cfg.scramble_en &= !scramble_disable;
+        cfg.ecc_en &= !ecc_disable;
+        /* TODO: also support OTP HW_INFO_CFG overrides */
```
Can you make an issue please

AlexJones0 replied:
Actually, I had misunderstood these OTP fields. I thought these were *additionally* applied on top of the HW_CFG_OVERRIDE registers (as a final separate override for the ROM).

Instead, looking at the ROM code, it seems like the ROM code reads these values to configure the HW_CFG_OVERRIDE registers (and then presumably lock them).

So, there is actually nothing more to implement here. I've removed the redundant comment.

AlexJones0 commented:
Thanks for the reviews!

There are some tests `//sw/device/tests:flash_ctrl_info_access_lc_{}` for the `dev`, `prod`, `prodend`, `rma` and `testunlocked0` states that at least test the lc_ctrl integration and SW qualification changes. Unfortunately due to some issues in the QEMU Bazel integration these are always passing (always boot into RMA state). Ideally this will be fixed on OpenTitan so we can actually test that QEMU is performing the correct behaviours. 

The hardware operations cannot be trivially tested until keymgr is implemented, so testing that will unfortunately have to wait a few more PRs.

AlexJones0 committed:
[ot] scripts/opentitan: clang-format: align BitFieldColonSpacing with checkpatch

`scripts/checkpatch.pl` wants there to be no spacing around colons in
BitFields, whereas `clang-format.yml` wants spaces on both sides. Change
`clang-format.yml` to match checkpatch, to better align with upstream
QEMU.

Applies the formatting changes to all existing files.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 committed:
[ot] scripts/opentitan: tests: Add passing boot_data test

The `boot_data_functest` is now passing consistently with these
new flash changes - likely the additional SW qualification changes
are sufficient to meet the requirements of this test, even if not fully
modelled / connected with the lc_ctrl.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 commented:
I've opened up a draft PR https://github.com/lowRISC/qemu/pull/178 that shows the changes to the flash_ctrl that will use the functionality introduced in this PR (keymgr secrets, lc_ctrl connections, a couple of other fixes).

The changes in this PR alone are sufficient to get `//sw/device/silicon_creator/lib:boot_data_functest_sim_qemu_rom_with_fake_keys` passing, and with the LC connections on top we also see `//sw/device/tests:flash_ctrl_info_access_lc_dev_sim_qemu_rom_with_fake_keys` and `//sw/device/tests:flash_ctrl_info_access_lc_test_unlocked0_sim_qemu_rom_with_fake_keys` passing as well.

luismarques commented:
Progresses https://github.com/lowRISC/opentitan/issues/28123
